src/main/scala/mesosphere/marathon/api/forwarder/AsyncUrlConnectionRequestForwarder.scala

```diff
@@ -0,0 +1,210 @@
+    .withConnectionSettings(connectionSettings)
+    .withMaxRetries(0)
+
+  class IgnoringHostnameVerifier extends javax.net.ssl.HostnameVerifier {
+    override def verify(hostname: String, sslSession: SSLSession): Boolean = true
+  }
+
+  private def cloneResponseStatusAndHeader(remote: HttpResponse, response: HttpServletResponse): Unit = {
+    response.setStatus(remote.status.intValue)
+
+    remote.headers.foreach {
+      case HttpHeader(name, value) =>
+        // Akka HTTP does not preserve case for headers.
+        if (name.equalsIgnoreCase("date"))
+          response.setHeader("Date", value)
+        else if (name.equalsIgnoreCase("vary"))
+          response.setHeader("Vary", value)
+        else if (name.equalsIgnoreCase(LeaderProxyFilter.HEADER_MARATHON_LEADER))
+          response.addHeader(LeaderProxyFilter.HEADER_MARATHON_LEADER, value)
+        else
+          response.addHeader(name, value)
+    }
+    response.addHeader(HEADER_VIA, viaValue)
+    Done
+  }
+
+  private def proxiedRequestHeaders(request: HttpServletRequest): Seq[HttpHeader] = {
+    val headers = Seq.newBuilder[HttpHeader]
+    // getHeaderNames() and getHeaders() are known to return null, see:
+    //http://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletRequest.html#getHeaders(java.lang.String)
+    val names = Option(request.getHeaderNames).map(_.asScala).getOrElse(Nil)
+    for {
+      name <- names
+      // Reverse proxies commonly filter these headers: connection, host.
+      //
+      // The connection header is removed since it may make sense to persist the connection
+      // for further requests even if this single client will stop using it.
+      if !name.equalsIgnoreCase("connection")
+
+      // The host header is used to choose the correct virtual host and should be set to the hostname
+      // of the URL for HTTP 1.1. Thus we do not preserve it, even though Marathon does not care.
+      if !name.equalsIgnoreCase("host")
+
+      // These headers cannot be set by Akka HTTP as raw headers
+      if !name.equalsIgnoreCase("content-type")
+      if !name.equalsIgnoreCase("content-length")
+      if !name.equalsIgnoreCase("user-agent")
+
+      headerValues <- Option(request.getHeaders(name))
+      headerValue <- headerValues.seq
+    } {
+      logger.debug(s"add RawHeader $name: $headerValue")
+      headers += RawHeader(name, headerValue)
+    }
+
+    headers += RawHeader(HEADER_VIA, viaValue)
+    val forwardedFor = Seq(
+      Option(request.getHeader(HEADER_FORWARDED_FOR)),
+      Option(request.getRemoteAddr)
+    ).flatten.mkString(",")
+    headers += RawHeader(HEADER_FORWARDED_FOR, forwardedFor)
+    Option(request.getHeader("User-Agent")).foreach { ua =>
+      headers += akka.http.scaladsl.model.headers.`User-Agent`(ua)
+    }
+    headers.result()
+  }
+
+  private def createAndConfigureConnection(asyncContext: AsyncContext, url: URL, clientRequest: HttpServletRequest): Future[HttpResponse] = {
```

Comment:
Nit: I think it would help to prefix the HTTP types. E.g. `jax.HttpServletRequest` and `akka.HttpResponse`. This would ease understanding the mappings.

Reply:
That's a great suggestion

Reply:
✅  on jax. There are too many packages for akka.